### PR TITLE
feat: PubSub for V6

### PR DESCRIPTION
**Pull request by stocaaro**
#### Description of changes
Introduce the PubSub category to V6.
- Remove Provider/Registry from PubSub
- Add import endpoints for `mqtt` and `iot` (by default export `iot` at the top)
- Migrate to RxJS
- Migrate to V6 `import`s

#### Usage example
```ts
import { Amplify } from 'aws-amplify';
import { PubSub } from '@aws-amplify/pubsub';
import config from './aws-exports';
Amplify.configure(config);

export function ExampleSubscriptionComponent() {
    const [messageCount, setMessageCount] = useState<number>(0);

    let sub;
    let pubsub;
    useEffect(() => {
        pubsub = new PubSub({ 
            region: 'us-east-1', 
            endpoint: 'wss://xxxxxxxxxxxxx.iot.us-east-1.amazonaws.com/mqtt'
            })
        sub = pubsub.subscribe('exampleTopic', (_msg) => {
            setMessageCount(messageCount + 1)
        });
        return () => sub?.unsubscribe?.();
    }, [])

    return <>{messageCount} - 
    <Button onClick={pubsub?.publish?.('exampleTopic', "Test message")}>
        Send Message
    </Button></>
}
```

#### Description of how you validated changes
- Tested by [porting a test app](https://github.com/stocaaro/pubsub-reconnect-sample/pull/1/files)
- [Migrated the reconnect test app](https://github.com/aws-amplify/amplify-js-samples-staging/pull/632) and got it [working against this change](https://github.com/stocaaro/amplify-js/actions/runs/6497600028/job/17648908277)

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/pubsub/src/Providers/AWSIot.ts [hunk 1]
```diff
@@ -1,41 +1,43 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
-import { MqttOverWSProvider, MqttProviderOptions } from './MqttOverWSProvider';
-import { Signer, Credentials } from '@aws-amplify/core';
-
+import { MqttOverWS, MqttOptions } from './MqttOverWS';
+import { Signer } from '@aws-amplify/core/internals/utils';
```
Comment (AllanZhengYP) [inline packages/pubsub/src/Providers/AWSIot.ts:4]: We are going to remove `Signer` class soon. Can you use `preSignUrl` instead? Signer was intended for backwards compabibility over functional `presignUrl`;


```suggestion
import { presignUrl } from '@aws-amplify/core/internals/aws-client-utils';
```

Comment (stocaaro) [reply packages/pubsub/src/Providers/AWSIot.ts:4]: I made a run at the requested migration. There aren't examples of this and Signer is doing a lot of enrichment / data cleaning that isn't done by presignUrl.  Can you provide some clarification / examples of how this migration is supposed to go?

Comment (stocaaro) [reply packages/pubsub/src/Providers/AWSIot.ts:4]: We talked this over. The change to make this happen is more involved than makes sense to pull into this change.

**Review by AllanZhengYP — DISMISSED** [review]
Looks good to me!

File: packages/pubsub/src/Providers/MqttOverWS.ts [hunk 2]
```diff
@@ -1,17 +1,22 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
+
+// @ts-ignore
```
Comment (stocaaro) [inline packages/pubsub/src/Providers/MqttOverWS.ts:4]: With ts-strict on, we need an ignore to import Paho which is js only

**Review by stocaaro — COMMENTED** [review]
I've been over this change. It definitely covers re-introducing PubSub interfaces for V6 and removing a lot of interfacing that we don't want included in V6.

There may be some more dead code to hunt down, but that further cleanup fit better in an effort to move shared code with API out to core/internal/utils (or wherever). Lastly, this change doesn't include Hub types, which I have a separate task to circle back and include.

**Review by jimblanc — DISMISSED** [review]
Looks good, nice work!

**Review by jimblanc — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]

**Pull request merged**